### PR TITLE
fix(job): avoid error when job is moved when processing

### PR DESCRIPTION
**Pull request by roggervalf**
fixes https://github.com/taskforcesh/bullmq/issues/1343

### CODE REVIEW

**Comment by loris** [comment]
Any update on this one? @manast

**Comment by manast** [comment]
@loris this PR does not fix the issue unfortunatelly.

**Review by manast — APPROVED** [review]
Crazy, I thought this change was already in 2.0

**Review by manast — CHANGES_REQUESTED** [review]
We need some documentation on the new arguments and behaviour.

**Comment by manast** [comment]
Ok, so a couple of things:
- I think the autoComplete should be a job setting, not an argument for delay. The reason is that we may want to use this feature for other things. For example, rate-limiter, if you detect in the process function that you want this job to rate limit, then you can just explicitly activate the limit, move the job back to wait and then set the autoComplete to false.
- We need to explain what the autoComplete do in the documentation (simply that by default jobs complete as soon as they finalize, but you can disable this behaviour by disabling autoComplete, which is useful for doing this and that. :).
- I do not think the autoComplete setting needs to be stored in the job persistently, it could just be a property of the Job instance.
-

**Comment by neighbaa** [comment]
This looks like a great feature and I'm looking forward to this update, but I noticed something - couldn't this pattern also affect or break backoff / retries?
Presumably, when the job resumes it will increase attemptsMade - if the job then throws an error, wouldn't the backoff function use the same number?

If that's the case, I guess attemptsMade could possibly be decremented whenever the job is moved while processing. Should that also be done manually, or should it happen by default if you use autoComplete?

If that's not the case, apologies for an unhelpful comment.

Cheers

File: docs/gitbook/patterns/process-step-jobs.md [hunk 1]
```diff
@@ -153,6 +159,7 @@ const worker = new Worker(
             step = Step.Finish;
             return Step.Finish;
           } else {
+            job.autoComplete = true; // prevent missing lock error
```
[the remark this answers is not in the document]
Comment (manast) [reply docs/gitbook/patterns/process-step-jobs.md:162]: I think you are right, the default must be auto-complete which is the current behavior.

File: src/classes/job.ts [hunk 2]
```diff
@@ -71,6 +71,11 @@ export class Job<
   ReturnType = any,
   NameType extends string = string,
 > {
+  /**
+   * By default jobs complete as soon as they finalize, but you can disable this behaviour by enabling autoComplete.
+   */
+  autoComplete: boolean = false;
```
[the remark this answers is not in the document]
Comment (manast) [reply src/classes/job.ts:77]: I agree, it should be true here.

**Comment by manast** [comment]
@neighbaa which use-case are you planing to implement if this PR gets merged?

**Comment by neighbaa** [comment]
> @neighbaa which use-case are you planing to implement if this PR gets merged?

I'll try and explain. Bearing in mind, I could be wrong about how attemptsMade works, or when it gets incremented.

I'm imagining this being a problem in a case where I conditionally create child jobs from a parent processor and call moveToWaitingChildren. Say the parent job has the exponential backoff strategy set with 10 attempts.
When the parent resumes, it is possible the condition will still be met to once again create new child jobs and call moveToWaitingChildren.

This could happen x number of times.
If x is 7, and then for some reason an error throws inside the parent processor, even if it's the first error, would the retry delay be set to 2 ^ 7 * delay ?
If x is already more than 10, would the job simply fail, or not retry?

I imagine it's the same for repeatedly delaying jobs on condition. (Possibly if I'm polling another service or something.)

I might be wrong, but I would have expected that even if the job has run 10 times already, the first error should follow the backoff strategy with attemptsMade = 1 and retry promptly.

Does this make sense?

File: tests/test_rate_limiter.ts [hunk 3]
```diff
@@ -215,52 +215,6 @@ describe('Rate Limiter', function () {
...
-
-    const numJobs = 4;
-    const startTime = new Date().getTime();
-
-    const worker = new Worker(queueName, async () => {}, {
-      connection,
-      limiter: {
-        max: 1,
-        duration: 1000,
-        workerDelay: true,
```
Comment (roggervalf) [inline tests/test_rate_limiter.ts:229]: workerDelay option does not exist in our code base

**Review by manast — APPROVED** [review]

**Pull request merged**